File: binary trees/binary_tree_diameter.py
# Binary Tree Diameter

# 1
class BinaryTree:
  def __init__(self, value, left=None, right=None):
    self.value = value
    self.left = left
    self.right = right


def binaryTreeDiameter(tree):
  ans = 0

  def dfs(root):
    nonlocal ans
    if not root:
      return 0

    left_h, right_h = dfs(root.left), dfs(root.right)

    ans = max(ans, left_h + right_h)

    return 1 + max(left_h, right_h)

  dfs(tree)

  return ans

File: binary trees/binary_tree_diameter.js
// #1

class BinaryTree {
constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
  }
}

function binaryTreeDiameter(tree) {
  // Write your code here.
  let ans = 0

  const height = root => {
    if (!root) {
      return 0
    }

    const leftHeight = height(root.left)
    const rightHeight = height(root.right)

    ans = Math.max(ans, leftHeight + rightHeight)

    return 1 + Math.max(leftHeight, rightHeight)
  }

  height(tree)

  return ans
}